### PR TITLE
NOTIF-484 Store email templates in the database

### PR DESCRIPTION
**Pull request by gwenneg**
This PR introduces the following classes:
- `Template` is used to store in the DB the content of the current HTML templates files.
- `InstantEmailTemplate` is used to store in the DB the link between an event type, a subject `Template` and a body `Template`.
- `AggregationEmailTemplate` is used to store in the DB the link between an application, a subscription type (currently only `DAILY`), a subject `Template` and a body `Template`.
- `TemplateRepository` is used to retrieve templates information from the DB. It is meant to replace both `EmailTemplateFactory` and the various `EmailTemplate` implementations that currently exist (`Policies`, `Advisor`, `Rhosak`...).

This PR does NOT contain:
- The `backend` module code that will be used in production to store/edit/delete the templates.
- The REST endpoints that will be used from the frontend.
- The migration code that will be needed to transform the current HTML files and Java logic into DB data.

This will all be done in subsequent PRs.

ℹ️ All the changes from this PR are disabled by default and can be enabled from configuration. Merging the PR should not affect in any way the templates rendering on prod.

### CODE REVIEW

File: engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java [hunk 1]
```diff
@@ -105,9 +120,14 @@ void postConstruct() {
         } else {
             Action action = event.getAction();
             final EmailTemplate template = emailTemplateFactory.get(action.getBundle(), action.getApplication());
-            final boolean shouldSaveAggregation = Arrays.stream(EmailSubscriptionType.values())
-                    .filter(emailSubscriptionType -> emailSubscriptionType != EmailSubscriptionType.INSTANT)
-                    .anyMatch(emailSubscriptionType -> template.isSupported(action.getEventType(), emailSubscriptionType));
+            boolean shouldSaveAggregation;
+            if (useTemplatesFromDb) {
+                shouldSaveAggregation = templateRepository.isSupported(action.getBundle(), action.getApplication());
```
Comment (gwenneg) [inline engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:125]: Temp, will change soon.

File: engine/src/main/java/com/redhat/cloud/notifications/models/ApplicationEmailTemplateId.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+@Embeddable
+public class ApplicationEmailTemplateId implements Serializable {
+
+    @NotNull
+    public UUID applicationId;
+
+    @NotNull
+    @Size(max = 50)
+    @Convert(converter = EmailSubscriptionTypeConverter.class)
+    public EmailSubscriptionType subscriptionType;
```
[the remark this answers is not in the document]
Comment (gwenneg) [reply engine/src/main/java/com/redhat/cloud/notifications/models/ApplicationEmailTemplateId.java:22]: Here's how I imagine things regarding the instant notification templates at application/bundle level:
- We would first look for an "instant" template at event type level in the DB.
- If not found, we would look for an "instant" template at application level.
- If not found, same thing at bundle level in the DB.

[the remark this answers is not in the document]
Comment (gwenneg) [reply engine/src/main/java/com/redhat/cloud/notifications/models/ApplicationEmailTemplateId.java:22]: Templates are not changing a lot, so we would cache the templates for 1 day to limit the number of SQL queries and clear the cache on template update.

[the remark this answers is not in the document]
Comment (gwenneg) [reply engine/src/main/java/com/redhat/cloud/notifications/models/ApplicationEmailTemplateId.java:22]: Better naming proposal:
- EventTypeInstantEmailTemplate (currently `EventTypeEmailTemplate`)
- ApplicationInstantEmailTemplate (fallback if nothing found at event type level, not created yet)
- BundleInstantEmailTemplate (fallback if nothing found at app level, not created yet)
- ApplicationDailyEmailTemplate (currenlty `ApplicationEmailTemplate`)
- BundleDailyEmailTemplate (fallback if nothing found at app level, not created yet)

[the remark this answers is not in the document]
Comment (gwenneg) [reply engine/src/main/java/com/redhat/cloud/notifications/models/ApplicationEmailTemplateId.java:22]: @pilhuhn You may have suggestions/ideas about this.

[the remark this answers is not in the document]
Comment (josejulio) [reply engine/src/main/java/com/redhat/cloud/notifications/models/ApplicationEmailTemplateId.java:22]: I'm not really sure if we want to add the fallback rules. 
I don't think we would like to fallback to bundle globals or even app globals templates. 

Right now the templates are tighly coupled with the payload, but nothing is enforcing that the data is being consumed correctly (inside the template). Qute provides a way to validate - but we would need to have a schema in place first.

I think is best if the apps select a specific template for each event (it could indeed be the same template, but they would select and test it manually).

Besides, a global template won't be user friendly or contain pertinent information IMO - I think it would be more useful when onboarding, and we can do something simpler like `if !templateFound && env != prod => use_onboarding_template`.

About new types, we might add a weekly subscription type and we would indeed support it only at the application level.
We can keep the `subscriptionType` param to keep track of this. 

Would like to hear other's thoughts about this.

[the remark this answers is not in the document]
Comment (gwenneg) [reply engine/src/main/java/com/redhat/cloud/notifications/models/ApplicationEmailTemplateId.java:22]: If fallbacks were implemented, there would certainly be a condition (boolean field) to enable/disable them depending on the context, and they would probably be disabled most of the time.

But maybe we don't need fallbacks at all. It's not in that PR anyway 😄

File: engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java [hunk 3]
```diff
@@ -105,9 +123,13 @@ void postConstruct() {
...
             Action action = event.getAction();
             final EmailTemplate template = emailTemplateFactory.get(action.getBundle(), action.getApplication());
-            final boolean shouldSaveAggregation = Arrays.stream(EmailSubscriptionType.values())
-                    .filter(emailSubscriptionType -> emailSubscriptionType != EmailSubscriptionType.INSTANT)
-                    .anyMatch(emailSubscriptionType -> template.isSupported(action.getEventType(), emailSubscriptionType));
+            boolean shouldSaveAggregation;
+            if (useTemplatesFromDb) {
+                shouldSaveAggregation = templateRepository.isEmailAggregationSupported(action.getBundle(), action.getApplication(), NON_INSTANT_SUBSCRIPTION_TYPES);
+            } else {
+                shouldSaveAggregation = NON_INSTANT_SUBSCRIPTION_TYPES.stream()
```
Comment (josejulio) [inline engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:130]: Not part of this PR, but i'm questioning now this mechanism.
Now that we have the event log we should probably take it from there.

Comment (gwenneg) [reply engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:130]: I want to make sure I understood your comment. Do you mean that we should use the payloads from the `Event#payload` rather than duplicating them into `email_aggregation`? If so, I agree with you 😃 The whole duplication followed by a purge process doesn't seem useful anymore since the DB contains the timestamp of the latest aggregation.

Comment (gwenneg) [reply engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:130]: We would gain in flexibility if we used `Event#payload`: each account could have a different aggregation time frame (1 day, 1 week, 2 hours...) with minimal code/DB changes.

Comment (josejulio) [reply engine/src/main/java/com/redhat/cloud/notifications/processors/email/EmailSubscriptionTypeProcessor.java:130]: > Do you mean that we should use the payloads from `Event#payload` rather than duplicating them into `email_aggregation`?

That's what I meant :D

**Comment by josejulio** [comment]
Github actions didn't run on this PR :(

**Comment by gwenneg** [comment]
I'll push again to trigger the actions.

**Comment by gwenneg** [comment]
GH finally accepted to run the actions 😃

**Review by josejulio — APPROVED** [review]
lgtm!
